open-craft/opencraft
# [SE-2869] Forbid registering domains/subdomains which would trick our DNS settings

This PR implements a domain/subdomain validation for custom and managed past/archived/active instances. The implementation uses three approaches to validate the domain:

1. If the domain starts with a reserved word (preview/studio/ecommerence/discovery), the validation will fail
2. If the subdomain starts with a reserved word (preview/studio/ecommerence/discovery), the validation will fail
3. If the domain is a subdomain of an existing instance's domain, the registration will fail
4. If the subdomain is already in our DNS settings (like haproxy.net.opencraft.com or ldp.opencraft.hosting), the validation will fail

**Dependencies**: None

**Screenshots**:

Requiring ack from the user before submission
![Screenshot 2020-09-18 at 13 35 12](https://user-images.githubusercontent.com/19173947/93594383-031fc800-f9b6-11ea-8e3c-6d6324643337.png)

(1) Domain starts with a reserved word
![Screenshot 2020-09-16 at 22 16 32](https://user-images.githubusercontent.com/19173947/93387959-4b31d400-f86a-11ea-9856-fd2bc8a87ffd.png)

(2) Subdomain starts with a reserved word
![Screenshot 2020-09-16 at 22 14 34](https://user-images.githubusercontent.com/19173947/93387874-2fc6c900-f86a-11ea-80bb-2c2672118d5a.png)
![Screenshot 2020-09-16 at 22 14 44](https://user-images.githubusercontent.com/19173947/93387885-30f7f600-f86a-11ea-9157-9eeab19ef242.png)
![Screenshot 2020-09-16 at 22 14 52](https://user-images.githubusercontent.com/19173947/93387891-31908c80-f86a-11ea-8194-22a52491f976.png)

(3) Subdomain of an existing instance's domain
![Screenshot 2020-09-16 at 22 11 03](https://user-images.githubusercontent.com/19173947/93387443-8bdd1d80-f869-11ea-8977-7572b99c731c.png)
![Screenshot 2020-09-16 at 22 11 40](https://user-images.githubusercontent.com/19173947/93387496-9d262a00-f869-11ea-9957-c23fbbe13fc1.png)

(4) Subdomain is already in our DNS settings
<img width="723" alt="Screenshot 2020-09-17 at 8 22 47" src="https://user-images.githubusercontent.com/19173947/93428028-1308b080-f8bf-11ea-9bdf-0a75867fae33.png">
<img width="725" alt="Screenshot 2020-09-17 at 8 22 55" src="https://user-images.githubusercontent.com/19173947/93428032-1439dd80-f8bf-11ea-8301-0e28fc0360df.png">

**Merge deadline**: None

**Testing instructions**:

1. Start the backend service
2. Start the frontend by executing `REACT_APP_INTERNAL_DOMAIN_NAME=".<YOUR_DOMAIN>" npm start` where `YOUR_DOMAIN` is equal to the `DEFAULT_INSTANCE_BASE_DOMAIN` and be aware of the `.` at the beginning.  
3. Navigate to `http://localhost:3000/registration/domain`
4. Try any of the scenarios shown on the screenshots above (please note that the `DEFAULT_INSTANCE_BASE_DOMAIN` defaults to `plebia.net` where only some DNS records are set)

**Author notes and concerns**:

* **BetaTestApplication's `BASE_DOMAIN` is set to the `DEFAULT_INSTANCE_BASE_DOMAIN`** as it seems that's the correct approach, since setting `opencraft.hosting` domain for every beta test app does not make sense if someone sets `DEFAULT_INSTANCE_BASE_DOMAIN` to something else (like `plebia.net`) and has no control over `opencraft.hosting` domain's DNS records.
* The scenario discussed in `test_external_domain_forbidden_domain_not_causing_issue` does not exists for subdomains
* In case we cannot access the `DEFAULT_INSTANCE_BASE_DOMAIN` DNS records, **we won't allow** registering users with the given subdomain **to make sure we are not overriding possibly existing DNS records** because of a network glitch or similar.
* Watch out for betatestapplication 0018 migration! (that's only a help text and validator error code change though)
* The domain accept text is **required for subdomains as well** on the registration page when not registering with custom domain (this could make sense in some cases)
* The js/ts client is updated (`npm run update-api-client`), hence some changes which may not be related completely to this change
* There is a really-really small chance that we will raise an error for the user if an OC member manually registers a subdomain on Gandi but that's not in the cache (since we have no callback from Gandi to OCIM)
